### ISSUE TITLE
[FEATURE] - Reactive support for Pulsar consumption

### ISSUE THREAD

**Issue by migroskub** [majusko/pulsar-java-spring-boot-starter#190]
**Is your feature request related to a problem? Please describe.**
I would like to use the library and use Spring with Pulsar, but I currently use Project Reactor with Spring, hence I would like to be able to consume Pulsar and use the `Flux` that it will return.

**Describe the solution you'd like**
A function in which I would be able to consume Pulsar and get `Flux` of messages, which I will be able to use with my reactive application.

**Describe alternatives you've considered**
A user can use `Mono.fromFuture` and implement the Pulsar-Project Reactor integration. An out-of-the-box solution would be better.

**Comment by lynch19** [comment]
+1

**Comment by lynch19** [comment]
This would be great for Project Reactor users, which are currently limited and need to implement such integration on their own.

**Comment by roger751** [comment]
+1

**Comment by majusko** [comment]
Ok, I see strong support :) I will have a look into this. It shouldn't be that complicated since the reactor for error stream is already there.

**Comment by thepaep** [comment]
Seems great. Looking forward to this too

**Comment by migroskub** [comment]
@majusko update us in case there is a need for help with contributions

**Comment by majusko** [comment]
@migroskub The implementation of reactor core support is done in a PR linked to this issue.

You should be able to create a bean that will handle subscriptions reactively:

```java
@Configuration
public class MyFluxConsumers {
    
    @Autowired
    private FluxConsumerFactory fluxConsumerFactory;

    @Bean
    public FluxConsumer<MyMsg> myFluxConsumer() {
        return fluxConsumerFactory.newConsumer(
            PulsarFluxConsumer.builder()
                .setTopic("flux-topic")
                .setConsumerName("flux-consumer")
                .setSubscriptionName("flux-subscription")
                .setClazz(MyMsg.class)
                .build());
    }
}
```

You are now able to subscribe to it:

```java
@Service
public class MyFluxConsumerService {
    
    @Autowired
    private FluxConsumer<MyMsg> myFluxConsumer;

    @EventListener(ApplicationReadyEvent.class)
    public void subscribe() {
        myFluxConsumer
            .asSimpleFlux()
            .subscribe(msg -> System.out.println(msg.getData()));
    }
}
```

In case you wish to acknowledge messages on your own. (Due to the asynchronous manner of reactive core, it's something people probably want to do more often). You can do this with little change in the configuration:

```java
PulsarFluxConsumer.builder()
    // ....
    .setSimple(false) // This is your required change in bean configuration class
    .build());
```

Subscription will then look like this:

```java
@Service
public class MyFluxConsumerService {
    
    @Autowired
    private FluxConsumer<MyMsg> myFluxConsumer;

    @EventListener(ApplicationReadyEvent.class)
    public void subscribe() {
        myFluxConsumer.asFlux()
            .subscribe(msg -> {
                try {
                    final MyMsg myMsg = (MyMsg) msg.getMessage().getValue();

                    System.out.println(myMsg.getData());

                    // you need to acknowledge the message manually on finished job
                    msg.getConsumer().acknowledge(msg.getMessage());
                } catch (PulsarClientException e) {
                    // you need to negatively acknowledge the message manually on failures
                    msg.getConsumer().negativeAcknowledge(msg.getMessage());
                }
            });
    }
}
```

Feel free to do a code review and let me know if you have any suggestions. I will merge and release a new version soon. (1-2 weeks).

**Comment by migroskub** [comment]
This is awesome!! Thank you

**Comment by majusko** [comment]
@migroskub I renamed the `setClazz ` to `setMessageClass `. Thanks for the feedback and for helping the library.

**Comment by majusko** [comment]
Reactive support has been released in the 1.1.0 minor update. I will implement an example of using it in an example project soon.
